flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriterState.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.kafka.sink;
+
+import javax.annotation.Nullable;
+
+import java.util.Properties;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class holds the necessary information to construct a new {@link FlinkKafkaInternalProducer}
+ * to continue processing of the {@link KafkaWriter} after a possible interruption i.e failure.
+ */
+class KafkaWriterState {
+
+    private final long producerId;
+    private final short epoch;
+
+    @Nullable private transient FlinkKafkaInternalProducer<byte[], byte[]> producer;
+    private final Properties kafkaProducerConfig;
+
+    public KafkaWriterState(
+            Properties kafkaProducerConfig, FlinkKafkaInternalProducer<byte[], byte[]> producer) {
+        this(kafkaProducerConfig, -1, (short) -1, producer);
+    }
+
+    public KafkaWriterState(Properties kafkaProducerConfig, long producerId, short epoch) {
+        this(kafkaProducerConfig, producerId, epoch, null);
+    }
+
+    public FlinkKafkaInternalProducer<byte[], byte[]> getProducer() {
+        if (producer == null) {
+            producer = new FlinkKafkaInternalProducer<>(kafkaProducerConfig);
+            producer.resumeTransaction(producerId, epoch);
+        }
+        return producer;
+    }
+
```

Comment:
Keep state dumb. Just return producerId and let using code constructor the producer.

Reply:
Hmm, I really like the abstraction that consumers only call `getProducer` and do not have to worry about the initialization but I agree a state object usually implies not much logic.

Reply:
You can simply introduce a wrapper around state that also manages the producer. In that way, we don't need to think about the additional things in the serializer.